sklearn/metrics/tests/test_pairwise.py

```diff
@@ -893,3 +896,37 @@ def test_check_preserve_type():
                                                    XB.astype(np.float))
     assert_equal(XA_checked.dtype, np.float)
     assert_equal(XB_checked.dtype, np.float)
+
+
+@pytest.mark.parametrize("n_jobs", [1, 2, -1])
+@pytest.mark.parametrize("metric", ["seuclidean", "mahalanobis"])
+@pytest.mark.parametrize("dist_function",
+                         [pairwise_distances, pairwise_distances_chunked])
+@pytest.mark.parametrize("y_is_x", [True, False], ids=["Y is X", "Y is not X"])
+def test_pairwise_distances_data_derived_params(n_jobs, metric, dist_function,
+                                                y_is_x):
+    # check that pairwise_distances give the same result in sequential and
+    # parallel, when metric has data-derived parameters.
+    with config_context(working_memory=0.1):  # to have more than 1 chunk
+        rng = np.random.RandomState(0)
+
+        X = rng.random_sample((1000, 10))
+
+        if y_is_x:
+            Y = X
+            if metric == "seuclidean":
+                params = {'V': np.var(X, axis=0, ddof=1)}
+            else:
+                params = {'VI': np.linalg.inv(np.cov(X.T)).T}
+        else:
+            Y = rng.random_sample((1000, 10))
+            if metric == "seuclidean":
+                params = {'V': np.var(np.vstack([X, Y]), axis=0, ddof=1)}
+            else:
+                params = {'VI': np.linalg.inv(np.cov(np.vstack([X, Y]).T)).T}
+
+        expected_dist = cdist(X, Y, metric=metric, **params)
```

Comment:
why do the params need to be passed explicitly. doesn't cdist calculate these by default?

Reply:
Yes but as I've been trying to explain, `pdist(X)` and `cdist(X,X)` disagree, since var is computed on X for pdist and on [X,X] for cdist.

This fix proposes the following:
- `pairwise_distances(X, Y!=X)` gives same output as `cdist(X,Y)`
- `pairwise_distances(X)` or `pairwise_distances(X,X)` gives same output as `pdist(X)` or equivalently as `cdist(X, X, V=var(X))`
- `pairwise_distances(X, Y==X)` (but Y is not X) gives same output as `cdist(X, X)`

Reply:
Note that if we want `pairwise_distances(X,X)` to give same output as `cdist(X, X)`, then `pairwise_distances(X,X) != pairwise_distances(X)` which felt bad.

Reply:
I'm not sure I follow entirely.
We have ``pairwise_distances(X, X) == pairwise_distances(X) != pairwise_distances(X, X.copy())`` right now, right?

And we're using ``var(X)`` in the first two and ``var([X, X])`` in the last one.
Is that not the same behavior as cdist?

Reply:
Ok. Shouldn't we then assert that in addition to explicitly passing params?

Reply:
So I think @amueller is asking that we test the equivalence of dist_function to cdist or pdist as appropriate, where we do not explicitly pass params. At the moment we are only comparing to when params are given.

Reply:
I had a comment in between yours which seems to have been deleted... Anyway, I added an assert to explicitly check the equivalence of pairwise distances with the appropriate pdist or cdist.

I'd like to point out the fact that this new assert is basically a scipy test, because it's just assert that pdist and cdist do agree if passed appropriate params. But I guess it doesn't hurt to add extra checks, since it was not that clear at first.

Reply:
It's also documenting behavior and ensuring that our understanding of the relative behavior is still correct.

Reply:
good point thx